### PR TITLE
tests: document nested suites

### PR DESCRIPTION
**Pull request by zyga**

### CODE REVIEW

File: tests/nested/classic/hotplug/task.yaml [hunk 1]
```diff
@@ -1,5 +1,19 @@
 summary: Create ubuntu classic image, install snapd and test hotplug feature
 
+details: |
+    Snapd contains a little-used subsystem for reacting to dynamic device
+    reconfiguration events, and in response maintain a set of system-provided
```
Comment (degville) [inline tests/nested/classic/hotplug/task.yaml:5]:
```suggestion
    reconfiguration events, and in response, maintains a set of system-provided
```

File: tests/nested/classic/hotplug/task.yaml [hunk 2]
```diff
@@ -1,5 +1,19 @@
...
+    slots for certain snapd interfaces.
+
+    One of the first interfaces to support this feature was the serial-port
+    interface, where snapd was able to detect serial ports that are not declared
```
Comment (degville) [inline tests/nested/classic/hotplug/task.yaml:9]:
```suggestion
    interface. Using this, snapd was able to detect serial ports that were not declared
```

File: tests/nested/classic/hotplug/task.yaml [hunk 3]
```diff
@@ -1,5 +1,19 @@
...
+    statically in the gadget snap, but instead were enumerated on startup or
+    discovered dynamically when attached to the system over the USB bus.
+
+    Testing this sort of interaction requires a nested virtual machine, where
```
Comment (degville) [inline tests/nested/classic/hotplug/task.yaml:13]:
```suggestion
    Simulating and testing this sort of interaction requires a nested virtual machine, where
```

File: tests/nested/core/core-revert/task.yaml [hunk 4]
```diff
@@ -1,5 +1,13 @@
 summary: core revert test
 
+details: |
+    Snapd used to change the way seccomp profiles are handled, which could cause
```
Comment (degville) [inline tests/nested/core/core-revert/task.yaml:4]:
```suggestion
    Older versions of snapd used to change the way seccomp profiles are handled, which could cause
```

File: tests/nested/core/core-revert/task.yaml [hunk 5]
```diff
@@ -1,5 +1,13 @@
...
+    problems when future snapd was reverted to old snapd, causing old profiles
```
Comment (degville) [inline tests/nested/core/core-revert/task.yaml:5]:
```suggestion
    problems when future snapd was reverted to old snapd, causing old profiles to
```

File: tests/nested/core/core-revert/task.yaml [hunk 6]
```diff
@@ -1,5 +1,13 @@
...
+    fail to parse and load. This was addressed with the introduction of system
```
Comment (degville) [inline tests/nested/core/core-revert/task.yaml:6]:
```suggestion
    fail to parse and load. This was addressed with the introduction of a system
```

File: tests/nested/core/core-revert/task.yaml [hunk 7]
```diff
@@ -1,5 +1,13 @@
...
+    key and eventually with the switch to pre-compiled seccomp profiles. This
+    test uses a virtual machine to ensure, that the core snap can be safely
```
Comment (degville) [inline tests/nested/core/core-revert/task.yaml:8]:
```suggestion
    test uses a virtual machine to ensure that the core snap can be safely
```

File: tests/nested/core/core20-kernel-failover/task.yaml [hunk 8]
```diff
@@ -2,7 +2,7 @@ summary: Check that a broken kernel snap automatically rolls itself back
 
 details: |
   Check that when it is triggered an installation of a broken kernel (there are 7 different
```
Comment (degville) [inline tests/nested/core/core20-kernel-failover/task.yaml:4]:
```suggestion
  Check the trigger for a broken kernel installation (there are 7 different
```

File: tests/nested/core/core20-kernel-failover/task.yaml [hunk 9]
```diff
@@ -2,7 +2,7 @@ summary: Check that a broken kernel snap automatically rolls itself back
...
-  kernel corruptions), the install change sinished with error and the initial kernel revision
+  kernel corruptions), the install change finished with error and the initial kernel revision
```
Comment (degville) [inline tests/nested/core/core20-kernel-failover/task.yaml:5]:
```suggestion
  kernel corruptions) adds an install change error and that the initial kernel revision
```

File: tests/nested/manual/cloud-init-never-used-not-vuln/task.yaml [hunk 10]
```diff
@@ -2,6 +2,12 @@ summary: |
     Test that cloud-init is no longer vulnerable on Ubuntu Core with the fix for
     CVE-2020-11933 in place.
 
+details: |
+    CVE-2020-11933 allowed anyone to present the device with cloud-info
+    meta-data, for example on a removable media, and reboot the machine to gain
```
Comment (degville) [inline tests/nested/manual/cloud-init-never-used-not-vuln/task.yaml:7]: is it alright to make this exploitable implementation detail potentially public?

File: tests/nested/manual/cmdline-remove-append/task.yaml [hunk 11]
```diff
@@ -1,5 +1,11 @@
 summary: Check that gadget.yaml can remove/add kernel command lines
 
+details: |
+  Snapd gadget snaps have influence over the boot process. Among others, they
+  can control the kernel command line - a critical resource in many ways. The
+  test verifies that an updated gadget snap can effectively add or remove kernel
+  command line arguments.
```
Comment (degville) [inline tests/nested/manual/cmdline-remove-append/task.yaml:7]:
```suggestion
  Gadget snaps have influence over the boot process, and subsequently, 
  can control the kernel command line - a critical resource in many ways. This
  test verifies that an updated gadget snap can effectively add or remove kernel
  command line arguments.
```

File: tests/nested/manual/core-early-config/task.yaml [hunk 12]
```diff
@@ -1,5 +1,10 @@
 summary: Test that config defaults are applied early when image is created.
 
+details: |
+    Snapd gadget snaps have influence over the default configuration of seeded
+    snaps. The test verifies that an a gadget snap can effectively apply default
```
Comment (degville) [inline tests/nested/manual/core-early-config/task.yaml:5]:
```suggestion
    snaps. The test verifies that a gadget snap can effectively apply default
```

File: tests/nested/manual/core-early-config/task.yaml [hunk 13]
```diff
@@ -1,5 +1,10 @@
...
+    configuration settings to the system that is booted for the first time.
```
Comment (degville) [inline tests/nested/manual/core-early-config/task.yaml:6]:
```suggestion
    configuration settings to a system that is booted for the first time.
```

File: tests/nested/manual/core20-new-snapd-does-not-break-old-initrd/task.yaml [hunk 14]
```diff
@@ -1,16 +1,23 @@
 summary: verify that new snapd's do not break old snap-bootstrap/kernel initrds
 
+details: |
+  Snapd secure boot story is complex as some of the elements that live in the
```
Comment (degville) [inline tests/nested/manual/core20-new-snapd-does-not-break-old-initrd/task.yaml:4]:
```suggestion
  The snapd secure boot implementation is complex, and some of the elements that live in the
```

File: tests/nested/manual/core20-new-snapd-does-not-break-old-initrd/task.yaml [hunk 15]
```diff
@@ -1,16 +1,23 @@
...
+  snapd.git repository end up as a part of the signed kernel package's initrd
+  file. This allows a skew to occur, where future snapd is booted with past
```
Comment (degville) [inline tests/nested/manual/core20-new-snapd-does-not-break-old-initrd/task.yaml:6]:
```suggestion
  file. This inadvertently allows a skew to occur, where a future snapd is booted with a past
```

File: tests/nested/manual/gadget-connections/task.yaml [hunk 16]
```diff
@@ -1,5 +1,10 @@
 summary: Check that connections from gadget are automatically connected
 
+details: |
+    Snapd gadget snaps have influence over the default connections of seeded
+    snaps. The test verifies that an a gadget snap can effectively connect
```
Comment (degville) [inline tests/nested/manual/gadget-connections/task.yaml:5]:
```suggestion
    snaps. The test verifies that a gadget snap can effectively connect
```

**Review by bboozzoo — APPROVED** [review]
Thanks

File: tests/nested/manual/devmode-snaps-can-run-other-snaps/task.yaml [hunk 17]
```diff
@@ -3,7 +3,7 @@ summary: |
 
 details: |
   For xenial the test covers running core and non-core based devmode snaps from core based strict snap and
```
Comment (degville) [inline tests/nested/manual/devmode-snaps-can-run-other-snaps/task.yaml:5]:
```suggestion
  For Xenial, the test covers running core and non-core based devmode snaps from core based strict snaps. 
```

File: tests/nested/manual/devmode-snaps-can-run-other-snaps/task.yaml [hunk 18]
```diff
@@ -3,7 +3,7 @@ summary: |
...
-  and for boinic the test covers running a non-core based devmode snap from a non-core based strict snap as
+  and for Bionic the test covers running a non-core based devmode snap from a non-core based strict snap as
```
Comment (degville) [inline tests/nested/manual/devmode-snaps-can-run-other-snaps/task.yaml:6]:
```suggestion
  For Bionic, the test covers running a non-core based devmode snap from a non-core based strict snap as
```

File: tests/nightly/sbuild/task.yaml [hunk 19]
```diff
@@ -1,5 +1,18 @@
 summary: Ensure snapd builds correctly in sbuild
 
+details: |
+    Debian package of snapd is sensitive to new dependencies that are not
```
Comment (degville) [inline tests/nightly/sbuild/task.yaml:4]:
```suggestion
    The Debian package of snapd is sensitive to new dependencies that are not
```

**Review by degville — COMMENTED** [review]
Amazing work here @zyga! Thank you so much for doing this. I think your approach is sound, and I've just left some non-critical suggestions.

File: tests/nested/manual/cloud-init-never-used-not-vuln/task.yaml (hunk 10 above)
Comment (sergiocazzolato) [reply tests/nested/manual/cloud-init-never-used-not-vuln/task.yaml:7]: this cve is already fixed in snapd, so it should be ok

**Review by sergiocazzolato — APPROVED** [review]
Nice, thanks

**Pull request merged**